src/azure-cli-core/azure/cli/core/translator/_decorators.py

```diff
@@ -0,0 +1,75 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from knack.arguments import CLIArgumentType
+from azure.cli.core.profiles import CustomResourceType
+from azure.cli.core.decorators import Completer
+
+
+# action
+def action_class(action_cls):
+    return action_cls
+
+
+def action_class_by_factory(factory):
+    return factory
+
+
+# arg_type
+def register_arg_type(register_name, overrides=None, **kwargs):  # pylint: disable=unused-argument
+    return CLIArgumentType(overrides=overrides, **kwargs)
+
+
+def arg_type_by_factory(factory):
+    return factory
+
+
+# client_factory
+def client_factory_func(func):
+    return func
```

Comment:
The outer function name can be retrieved using following [PEP 3155](https://www.python.org/dev/peps/pep-3155/#example-with-nested-functions):

```py
import inspect


def get_printer(msg):
    def printer():
        print(msg)
    return printer


p = get_printer("Hello")
print(p.__qualname__)
print(inspect.getclosurevars(p))
```

Output:

```
get_printer.<locals>.printer
ClosureVars(nonlocals={'msg': 'Hello'}, globals={}, builtins={'print': <built-in function print>}, unbound=set())
```

This can eliminate the necessity of these decorators.

Reply:
Thanks for your advice. This is a good solution to retrieve data from outer function. But I still prefer decorators because them build a layer between command definitions and linked functions in architecture level, which can not only solve questions above but also bring more flexibility.